src/base/utils/fs.h

```diff
@@ -33,12 +33,38 @@
  * Utility functions related to file system.
  */
 
+#include <iterator>
+
+#include <QByteArray>
 #include <QString>
 
+class QIODevice;
+
 namespace Utils
 {
     namespace Fs
     {
+        // A wrapper class that satisfy LegacyOutputIterator requirement
+        class FileOutputIterator
+            : public std::iterator<std::output_iterator_tag, void, void, void, void>
+        {
+        public:
+            explicit FileOutputIterator(QIODevice &device, const int bufferSize = (4 * 1024));
+            ~FileOutputIterator();
+
+            // mimic std::ostream_iterator behavior
+            FileOutputIterator &operator=(const char c);
+            // TODO: make these `constexpr` in C++17
+            FileOutputIterator &operator*();
+            FileOutputIterator &operator++();
+            FileOutputIterator &operator++(int);
+
+        private:
+            QIODevice &m_device;
+            QByteArray m_buffer;
```

Comment:
the fact that you store state inside the iterator means that you really never should copy it. You should make it non-copyable (i.e. `FileOutputIterator(FileOutputIterator const&) = delete;`). This will most likely cause the call to `std::copy()` that you highlighted, to fail. Because it wants to copy the iterator.

The way this is solved in the standard library is by `ostream_iterator` not keeping any state, the buffering is done in the `ostream` object itself. You could try that. Alternatively you could allocate the buffer in a `boost::shared_array<>` so copies of the iterator still use the same underlying buffer.

Reply:
in fact, isn't your compiler warning you that this class has a destructor but an implicitly defined copy constructor and move constructor?

Reply:
>the fact that you store state inside the iterator means that you really never should copy it. You should make it non-copyable 

Thanks for the hint, from https://en.cppreference.com/w/cpp/named_req/Iterator:
> The type It satisfies CopyConstructible, and
> The type It satisfies CopyAssignable, and 

This must be the culprit of the msvc errors.
However I think those are really unnecessary requirements for my usage here, of course I should make the class non-copyable, but that still won't make `std::copy()` happy.

>The way this is solved in the standard library is by ostream_iterator not keeping any state, the buffering is done in the ostream object itself. You could try that. 

The first version I wrote was stateless, it would just pass every thing down to `QIODevice::putChar()`, however a quick benchmark shows bad performance, I'll need to look into that again.